File: vault-audit-tools/src/commands/client_activity.rs
use crate::vault_api::{extract_data, should_skip_verify, VaultClient};
use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fs::File;
use std::io::Read;

#[derive(Debug, Deserialize)]
struct MountInfo {
    #[serde(rename = "type")]
    mount_type: Option<String>,
    accessor: Option<String>,
}

#[derive(Debug, Deserialize)]
struct ActivityRecord {
    client_id: String,
    client_type: Option<String>,
    mount_accessor: Option<String>,
    mount_path: Option<String>,
    mount_type: Option<String>,
    entity_alias_name: Option<String>,
}

#[derive(Debug, Deserialize)]
struct EntityMapping {
    display_name: String,
    #[allow(dead_code)]
    mount_path: String,
    #[allow(dead_code)]
    mount_accessor: String,
    #[allow(dead_code)]
    username: Option<String>,
    #[allow(dead_code)]
    login_count: usize,
    #[allow(dead_code)]
    first_seen: String,
    #[allow(dead_code)]
    last_seen: String,
}

#[derive(Debug, Serialize)]
struct MountActivity {
    mount: String,
    #[serde(rename = "type")]
    mount_type: String,
    accessor: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    role: Option<String>,
    total: usize,
    entity: usize,
    non_entity: usize,
}

fn format_number(n: usize) -> String {
    let s = n.to_string();
    let mut result = String::new();
    for (i, c) in s.chars().rev().enumerate() {
        if i > 0 && i % 3 == 0 {
            result.push(',');
        }
        result.push(c);
    }
    result.chars().rev().collect()
}

#[allow(clippy::too_many_arguments)]
pub async fn run(
    start_time: &str,
    end_time: &str,
    vault_addr: Option<&str>,
    vault_token: Option<&str>,
    insecure: bool,
    group_by_role: bool,
    entity_map_path: Option<&str>,
    output: Option<&str>,
) -> Result<()> {
    let skip_verify = should_skip_verify(insecure);
    let client = VaultClient::from_options(vault_addr, vault_token, skip_verify)?;

    eprintln!("=== Vault Client Activity Analysis ===");
    eprintln!("Vault Address: {}", client.addr());
    eprintln!("Time Window: {} to {}", start_time, end_time);
    if skip_verify {
        eprintln!("⚠️  TLS certificate verification is DISABLED");
    }
    eprintln!();

    // Load entity mappings if provided
    let entity_map: Option<HashMap<String, EntityMapping>> = if let Some(path) = entity_map_path {
        eprintln!("Loading entity mappings from: {}", path);
        let mut file = File::open(path)
            .with_context(|| format!("Failed to open entity map file: {}", path))?;
        let mut contents = String::new();
        file.read_to_string(&mut contents)?;
        let map: HashMap<String, EntityMapping> = serde_json::from_str(&contents)
            .with_context(|| format!("Failed to parse entity map JSON: {}", path))?;
        eprintln!("Loaded {} entity mappings", map.len());
        Some(map)
    } else {
        None
    };

    // Build mount lookup map
    eprintln!("Fetching mount information...");
    let mount_map = fetch_mount_map(&client).await?;
    eprintln!("Found {} mounts", mount_map.len());

    // Fetch activity export
    eprintln!("Fetching client activity data...");
    let export_path = format!(
        "/v1/sys/internal/counters/activity/export?start_time={}&end_time={}&format=json",
        start_time, end_time
    );

    let export_text = client.get_text(&export_path).await?;

    // Parse NDJSON (newline-delimited JSON) or regular JSON
    let records: Vec<ActivityRecord> = if export_text.trim().starts_with('[') {
        // Regular JSON array
        serde_json::from_str(&export_text)?
    } else {
        // NDJSON - parse line by line
        export_text
            .lines()
            .filter(|line| !line.trim().is_empty())
            .filter_map(|line| serde_json::from_str(line).ok())
            .collect()
    };

    if records.is_empty() {
        eprintln!("No activity data found for the specified time range.");
        return Ok(());
    }

    eprintln!(
        "Processing {} activity records...",
        format_number(records.len())
    );

    // Group by mount and count unique clients
    let mut mount_activities: HashMap<String, MountActivityData> = HashMap::new();

    for record in &records {
        let accessor = record
            .mount_accessor
            .as_deref()
            .unwrap_or("unknown")
            .to_string();

        let (mount_path, mount_type) = if let Some(info) = mount_map.get(&accessor) {
            (info.0.clone(), info.1.clone())
        } else {
            (
                record
                    .mount_path
                    .clone()
                    .unwrap_or_else(|| "unknown".to_string()),
                record
                    .mount_type
                    .clone()
                    .unwrap_or_else(|| "unknown".to_string()),
            )
        };

        // Extract role/appcode if grouping by role
        let role = if group_by_role {
            // Try entity_alias_name from export first (Vault 1.20+)
            if let Some(alias_name) = &record.entity_alias_name {
                Some(alias_name.clone())
            } else if let Some(ref entity_map) = entity_map {
                // Fallback to entity map (Vault 1.16 or when alias_name is missing)
                entity_map
                    .get(&record.client_id)
                    .map(|e| e.display_name.clone())
            } else {
                None
            }
        } else {
            None
        };

        // Create unique key based on grouping mode
        let key = if group_by_role {
            format!(
                "{}|{}|{}|{}",
                mount_path,
                mount_type,
                accessor,
                role.as_deref().unwrap_or("unknown")
            )
        } else {
            format!("{}|{}|{}", mount_path, mount_type, accessor)
        };

        let activity = mount_activities
            .entry(key)
            .or_insert_with(|| MountActivityData {
                mount: mount_path,
                mount_type,
                accessor,
                role: role.clone(),
                total_clients: std::collections::HashSet::new(),
                entity_clients: std::collections::HashSet::new(),
                non_entity_clients: std::collections::HashSet::new(),
            });

        activity.total_clients.insert(record.client_id.clone());

        if record.client_type.as_deref() == Some("entity") {
            activity.entity_clients.insert(record.client_id.clone());
        } else {
            activity.non_entity_clients.insert(record.client_id.clone());
        }
    }

    // Convert to output format
    let mut results: Vec<MountActivity> = mount_activities
        .into_values()
        .map(|data| {
            // Concatenate mount + role for the mount field if role exists
            let mount_display = if let Some(ref role) = data.role {
                format!("{}{}", data.mount, role)
            } else {
                data.mount.clone()
            };

            MountActivity {
                mount: mount_display,
                mount_type: data.mount_type,
                accessor: data.accessor,
                role: None, // Don't include role as separate field anymore
                total: data.total_clients.len(),
                entity: data.entity_clients.len(),
                non_entity: data.non_entity_clients.len(),
            }
        })
        .collect();

    // Sort by mount path
    results.sort_by(|a, b| a.mount.cmp(&b.mount));

    // Calculate totals
    let total_clients: usize = results.iter().map(|r| r.total).sum();
    let total_entity: usize = results.iter().map(|r| r.entity).sum();
    let total_non_entity: usize = results.iter().map(|r| r.non_entity).sum();

    eprintln!();
    eprintln!("=== Summary ===");
    eprintln!("Total Clients: {}", format_number(total_clients));
    eprintln!("  Entity Clients: {}", format_number(total_entity));
    eprintln!("  Non-Entity Clients: {}", format_number(total_non_entity));
    eprintln!("Mounts Analyzed: {}", results.len());
    eprintln!();

    // Output results
    if let Some(output_path) = output {
        let file = File::create(output_path)
            .with_context(|| format!("Failed to create output file: {}", output_path))?;
        let mut writer = csv::Writer::from_writer(file);

        writer.write_record(["mount", "type", "accessor", "total", "entity", "non_entity"])?;
        for result in &results {
            writer.write_record([
                &result.mount,
                &result.mount_type,
                &result.accessor,
                &result.total.to_string(),
                &result.entity.to_string(),
                &result.non_entity.to_string(),
            ])?;
        }

        writer.flush()?;
        eprintln!("CSV written to: {}", output_path);
    } else {
        // JSON output to stdout
        println!("{}", serde_json::to_string_pretty(&results)?);
    }

    Ok(())
}

#[derive(Debug)]
struct MountActivityData {
    mount: String,
    mount_type: String,
    accessor: String,
    role: Option<String>,
    total_clients: std::collections::HashSet<String>,
    entity_clients: std::collections::HashSet<String>,
    non_entity_clients: std::collections::HashSet<String>,
}

async fn fetch_mount_map(client: &VaultClient) -> Result<HashMap<String, (String, String)>> {
    let mut map = HashMap::new();

    // Try /sys/mounts (secret engines)
    if let Ok(mounts_data) = client.get_json("/v1/sys/mounts").await {
        if let Ok(mounts) = extract_data::<HashMap<String, MountInfo>>(mounts_data) {
            for (path, info) in mounts {
                if let Some(accessor) = info.accessor {
                    map.insert(
                        accessor,
                        (
                            path,
                            info.mount_type.unwrap_or_else(|| "unknown".to_string()),
                        ),
                    );
                }
            }
        }
    }

    // Try /sys/auth (auth methods)
    if let Ok(auth_data) = client.get_json("/v1/sys/auth").await {
        if let Ok(auths) = extract_data::<HashMap<String, MountInfo>>(auth_data) {
            for (path, info) in auths {
                if let Some(accessor) = info.accessor {
                    map.insert(
                        accessor,
                        (
                            path,
                            info.mount_type.unwrap_or_else(|| "unknown".to_string()),
                        ),
                    );
                }
            }
        }
    }

    Ok(map)
}

File: internal/commands/client_activity.go
// Package commands contains executable command implementations.
package commands

import (
	"encoding/csv"
	"encoding/json"
	"fmt"
	"os"
	"sort"
	"strings"

	"github.com/trenner1/hashicorp-vault-audit-analysis/internal/utils"
	"github.com/trenner1/hashicorp-vault-audit-analysis/internal/vault"
)

// ActivityRecord represents a single activity record from Vault API.
type ActivityRecord struct {
	ClientID         string `json:"client_id"`
	ClientType       string `json:"client_type"`
	MountAccessor    string `json:"mount_accessor"`
	MountPath        string `json:"mount_path"`
	MountType        string `json:"mount_type"`
	EntityAliasName  string `json:"entity_alias_name"`
}

// EntityMapping represents entity metadata from the entity map.
// MountActivity represents aggregated activity for a mount.
type MountActivity struct {
	Mount    string `json:"mount"`
	Type     string `json:"type"`
	Accessor string `json:"accessor"`
	Total    int    `json:"total"`
	Entity   int    `json:"entity"`
	NonEntity int   `json:"non_entity"`
}

// mountActivityData tracks clients for a mount during aggregation.
type mountActivityData struct {
	mount          string
	mountType      string
	accessor       string
	role           string
	totalClients   map[string]bool
	entityClients  map[string]bool
	nonEntityClients map[string]bool
}

// RunClientActivity queries Vault API for client activity and outputs results.
//
// Queries /v1/sys/internal/counters/activity/export to fetch client usage metrics,
// then outputs CSV or JSON with client counts grouped by mount.
func RunClientActivity(
	start, end string,
	vaultAddr, vaultToken, vaultNamespace *string,
	insecure bool,
	groupByRole bool,
	entityMap *string,
	output *string,
) error {
	// Build Vault client options
	opts := vault.Options{SkipVerify: insecure}
	if vaultAddr != nil {
		opts.Addr = *vaultAddr
	}
	if vaultToken != nil {
		opts.Token = *vaultToken
	}
	if vaultNamespace != nil {
		opts.Namespace = *vaultNamespace
	}

	client, err := vault.NewFromOptions(opts)
	if err != nil {
		return fmt.Errorf("create vault client: %w", err)
	}

	fmt.Fprintf(os.Stderr, "=== Vault Client Activity Analysis ===\n")
	fmt.Fprintf(os.Stderr, "Vault Address: %s\n", client.Addr())
	fmt.Fprintf(os.Stderr, "Time Window: %s to %s\n", start, end)
	if insecure {
		fmt.Fprintf(os.Stderr, "⚠️  TLS certificate verification is DISABLED\n")
	}
	fmt.Fprintf(os.Stderr, "\n")

	// Load entity mappings if provided
	var entityMapData map[string]EntityMapping
	if entityMap != nil && *entityMap != "" {
		fmt.Fprintf(os.Stderr, "Loading entity mappings from: %s\n", *entityMap)
		data, err := loadEntityMapJSON(*entityMap)
		if err != nil {
			return fmt.Errorf("load entity map: %w", err)
		}
		entityMapData = data
		fmt.Fprintf(os.Stderr, "Loaded %d entity mappings\n", len(entityMapData))
	}

	// Build mount lookup map
	fmt.Fprintf(os.Stderr, "Fetching mount information...\n")
	mountMap, err := fetchMountMap(client)
	if err != nil {
		return fmt.Errorf("fetch mount map: %w", err)
	}
	fmt.Fprintf(os.Stderr, "Found %d mounts\n", len(mountMap))

	// Fetch activity export
	fmt.Fprintf(os.Stderr, "Fetching client activity data...\n")
	exportPath := fmt.Sprintf("/v1/sys/internal/counters/activity/export?start_time=%s&end_time=%s", start, end)

	exportText, err := client.GetRaw(exportPath)
	if err != nil {
		return fmt.Errorf("get activity export: %w", err)
	}

	// Parse NDJSON (newline-delimited JSON)
	var records []ActivityRecord
	for _, line := range strings.Split(exportText, "\n") {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		var record ActivityRecord
		if err := json.Unmarshal([]byte(line), &record); err != nil {
			continue
		}
		records = append(records, record)
	}

	if len(records) == 0 {
		fmt.Fprintf(os.Stderr, "No activity data found for the specified time range.\n")
		return nil
	}

	fmt.Fprintf(os.Stderr, "Processing %s activity records...\n", utils.FormatNumber(len(records)))

	// Group by mount and count unique clients
	mountActivities := make(map[string]*mountActivityData)

	for _, record := range records {
		accessor := record.MountAccessor
		if accessor == "" {
			accessor = "unknown"
		}

		// Look up mount info
		var mountPath, mountType string
		if info, ok := mountMap[accessor]; ok {
			mountPath = info[0]
			mountType = info[1]
		} else {
			mountPath = record.MountPath
			if mountPath == "" {
				mountPath = "unknown"
			}
			mountType = record.MountType
			if mountType == "" {
				mountType = "unknown"
			}
		}

		// Extract role/appcode if grouping by role
		var role string
		if groupByRole {
			if record.EntityAliasName != "" {
				role = record.EntityAliasName
			} else if len(entityMapData) > 0 {
				if mapping, ok := entityMapData[record.ClientID]; ok {
					role = mapping.DisplayName
				}
			}
		}

		// Create unique key based on grouping mode
		var key string
		if groupByRole {
			if role == "" {
				role = "unknown"
			}
			key = fmt.Sprintf("%s|%s|%s|%s", mountPath, mountType, accessor, role)
		} else {
			key = fmt.Sprintf("%s|%s|%s", mountPath, mountType, accessor)
		}

		// Get or create activity entry
		if _, ok := mountActivities[key]; !ok {
			mountActivities[key] = &mountActivityData{
				mount:             mountPath,
				mountType:         mountType,
				accessor:          accessor,
				role:              role,
				totalClients:      make(map[string]bool),
				entityClients:     make(map[string]bool),
				nonEntityClients:  make(map[string]bool),
			}
		}

		activity := mountActivities[key]
		activity.totalClients[record.ClientID] = true

		if record.ClientType == "entity" {
			activity.entityClients[record.ClientID] = true
		} else {
			activity.nonEntityClients[record.ClientID] = true
		}
	}

	// Convert to output format
	var results []MountActivity
	for _, data := range mountActivities {
		// Concatenate mount + role if grouping by role
		mountDisplay := data.mount
		if groupByRole && data.role != "" {
			mountDisplay = data.mount + " / " + data.role
		}

		results = append(results, MountActivity{
			Mount:     mountDisplay,
			Type:      data.mountType,
			Accessor:  data.accessor,
			Total:     len(data.totalClients),
			Entity:    len(data.entityClients),
			NonEntity: len(data.nonEntityClients),
		})
	}

	// Sort by mount path
	sort.Slice(results, func(i, j int) bool {
		return results[i].Mount < results[j].Mount
	})

	// Calculate totals
	totalClients := 0
	totalEntity := 0
	totalNonEntity := 0
	for _, r := range results {
		totalClients += r.Total
		totalEntity += r.Entity
		totalNonEntity += r.NonEntity
	}

	fmt.Fprintf(os.Stderr, "\n")
	fmt.Fprintf(os.Stderr, "=== Summary ===\n")
	fmt.Fprintf(os.Stderr, "Total Clients: %s\n", utils.FormatNumber(totalClients))
	fmt.Fprintf(os.Stderr, "  Entity Clients: %s\n", utils.FormatNumber(totalEntity))
	fmt.Fprintf(os.Stderr, "  Non-Entity Clients: %s\n", utils.FormatNumber(totalNonEntity))
	fmt.Fprintf(os.Stderr, "Mounts Analyzed: %d\n", len(results))
	fmt.Fprintf(os.Stderr, "\n")

	// Output results
	if output != nil && *output != "" {
		f, err := os.Create(*output)
		if err != nil {
			return fmt.Errorf("create output file: %w", err)
		}
		defer f.Close()

		w := csv.NewWriter(f)
		w.Write([]string{"mount", "type", "accessor", "total", "entity", "non_entity"})
		for _, r := range results {
			w.Write([]string{
				r.Mount,
				r.Type,
				r.Accessor,
				fmt.Sprintf("%d", r.Total),
				fmt.Sprintf("%d", r.Entity),
				fmt.Sprintf("%d", r.NonEntity),
			})
		}
		w.Flush()
		if err := w.Error(); err != nil {
			return fmt.Errorf("write csv: %w", err)
		}
		fmt.Fprintf(os.Stderr, "CSV written to: %s\n", *output)
	} else {
		// JSON output to stdout
		data, err := json.MarshalIndent(results, "", "  ")
		if err != nil {
			return fmt.Errorf("marshal json: %w", err)
		}
		fmt.Println(string(data))
	}

	return nil
}

// fetchMountMap builds a map of mount accessor to (path, type).
func fetchMountMap(client *vault.Client) (map[string][2]string, error) {
	mountMap := make(map[string][2]string)

	// Try /sys/mounts (secret engines)
	var mountsResp map[string]interface{}
	if err := client.Get("/v1/sys/mounts", &mountsResp); err == nil {
		if dataRaw, ok := mountsResp["data"]; ok {
			if data, ok := dataRaw.(map[string]interface{}); ok {
				for path, infoRaw := range data {
					if info, ok := infoRaw.(map[string]interface{}); ok {
						if accessor, ok := info["accessor"].(string); ok {
							mountType := "unknown"
							if t, ok := info["type"].(string); ok {
								mountType = t
							}
							mountMap[accessor] = [2]string{path, mountType}
						}
					}
				}
			}
		}
	}

	// Try /sys/auth (auth methods)
	var authResp map[string]interface{}
	if err := client.Get("/v1/sys/auth", &authResp); err == nil {
		if dataRaw, ok := authResp["data"]; ok {
			if data, ok := dataRaw.(map[string]interface{}); ok {
				for path, infoRaw := range data {
					if info, ok := infoRaw.(map[string]interface{}); ok {
						if accessor, ok := info["accessor"].(string); ok {
							mountType := "unknown"
							if t, ok := info["type"].(string); ok {
								mountType = t
							}
							mountMap[accessor] = [2]string{path, mountType}
						}
					}
				}
			}
		}
	}

	return mountMap, nil
}

// loadEntityMapJSON loads entity mappings from a JSON file.
func loadEntityMapJSON(path string) (map[string]EntityMapping, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}

	var result map[string]EntityMapping
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, err
	}

	return result, nil
}
